### PR TITLE
Hello Docker

### PR DESCRIPTION
**Pull request by zaneselvans**
Build Docker images appropriate for running the nightly PUDL data builds.

### NOTE: before merging
* Turn the `tox-pytest` GitHub Action back on. It's disabled for the time being since I am pushing a gazillion times to test the new `docker-build-push` action.
* Turn the `clear-nb-outputs` and `unit-tests` pre-commit hooks back on. I disabled them since they take a few seconds to run and aren't relevant to getting the Docker builds working.

### Changes:
* Added the `hadolint` Dockerfile linter to our pre-commit hooks.
* Created a new Dockerfile, currently in the `docker` directory. It uses the `condaforge/mambaforge` image as a base, creates a non-root `catalyst` user, and creates a software environment that includes PUDL and all of its dependencies.
* Created a new `docker-build-push` workflow that builds the new Docker image, pushes it to `catalystcoop/pudl-etl`, and tags it with the tags / refs obtained directly from git using `docker/metadata-action`.
* Consolidated all of the old Dockerfile / release stuff from Jan's work under the `docker` directory, to be culled once we have the new setup working well.

### CODE REVIEW

**Comment by zaneselvans** [comment]
The pre-commit hooks are going to keep failing until we remove the old Dockerfiles, but I want to keep them around as a referenced until after I get #1606 finished.

File: .github/workflows/docker-build-push.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        uses: docker/metadata-action@v3.8.0
+        with:
+          images: catalystcoop/pudl-etl
+          flavor: |
+            latest=auto
+          tags: |
+            type=ref,event=branch
+            type=ref,event=tag
+
+      - name: Set up Docker Buildx
```
Comment (bendnorman) [inline .github/workflows/docker-build-push.yml:23]: So this action makes buildx available for the "Build image and push to Docker Hub" step?

**Review by bendnorman — COMMENTED** [review]
Seems pretty straightforward to me other than the conda env business.

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    conda clean -afy && \
+    conda run --prefix /home/catalyst/env pip install -e './[dev,doc,test]' && \
+    # Create data input/output directories
+    mkdir /home/catalyst/pudl_in && mkdir /home/catalyst/pudl_out && \
+    # Run the PUDL setup script so we know where to read and write data
+    conda run --prefix /home/catalyst/env \
+        pudl_setup --pudl_in /home/catalyst/pudl_in --pudl_out /home/catalyst/pudl_out
+
+# Run the unit tests:
+CMD ["conda", "run", "--prefix", "/home/catalyst/env", "pytest", "test/unit"]
```
[the remark this answers is not in the document]
Comment (zaneselvans) [reply docker/Dockerfile:31]: Yeah, this is just a placeholder that does something which demonstrates the environment is working and PUDL installed.

[the remark this answers is not in the document]
Comment (zaneselvans) [reply docker/Dockerfile:31]: And if we run it using `docker compose` we'll probably inject a command directly which overrides this one.

File: .github/workflows/docker-build-push.yml (hunk 1 above)
Comment (zaneselvans) [reply .github/workflows/docker-build-push.yml:23]: It does that, but I think does other magical things too, or maybe buildx does those things on its own -- like creating a Docker build context that isn't the surrounding directory, and instead is derived directly from the git repository (which we then have to override below, to get access to the .git/ folder for `setuptools_scm`)

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -130,5 +136,5 @@ ci:
   autoupdate_branch: dev
   autoupdate_commit_msg: '[pre-commit.ci] pre-commit autoupdate'
   autoupdate_schedule: weekly
-  skip: [unit-tests, nb-output-clear]
+#  skip: [unit-tests, nb-output-clear]
```
[the remark this answers is not in the document]
Comment (zaneselvans) [reply .pre-commit-config.yaml:139]: See the comment on the PR. They're just annoyingly slow and not related to what I'm changing right now.

**Comment by zaneselvans** [comment]
The `gce-deploy` and `hello-docker` branches are currently the same, so I'm going to close this PR and delete this branch.

**Pull request closed without merging**